# Step 01: Regression

In [1]:
import pandas as pd
df1 = pd.read_csv('weight-height.csv')
df1.head()

Gender     Height      Weight
0   Male  73.847017  241.893563
1   Male  68.781904  162.310473
2   Male  74.110105  212.740856
3   Male  71.730978  220.042470
4   Male  69.881796  206.349801

In [2]:
df2 = df1.copy()

2. Separate x(Gender, Height) and y (y=Weight)
3. Train = 70%, Test = 30%
4. Apply Linear Regression
5. Evaluate the Model (Testing and training Accuracy, MSE for testing)
6. Apply KNN Regressor: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html
7. Evaluate the Model (Testing and training Accuracy, MSE for testing)
8. Compare KNN & Linear regression with the KNN model and Linear regression as well.


Work from here..

# Separate X and Y

In [3]:
x= df1.iloc[:,:-1]
x.head()

Gender     Height
0   Male  73.847017
1   Male  68.781904
2   Male  74.110105
3   Male  71.730978
4   Male  69.881796

In [4]:
y=df1['Weight']
y.head()

0    241.893563
1    162.310473
2    212.740856
3    220.042470
4    206.349801
Name: Weight, dtype: float64

In [5]:
#x contain String(Gender) ,,,so it will be encodded

# One hot Encoding

In [5]:
gender = pd.get_dummies(x['Gender'])
gender.head()

Female  Male
0       0     1
1       0     1
2       0     1
3       0     1
4       0     1

In [6]:
x=x.drop(['Gender'], axis=1)
x.head()

Height
0  73.847017
1  68.781904
2  74.110105
3  71.730978
4  69.881796

In [7]:
x=pd.concat([x,gender], axis=1)
x.head()

Height  Female  Male
0  73.847017       0     1
1  68.781904       0     1
2  74.110105       0     1
3  71.730978       0     1
4  69.881796       0     1

In [8]:
from sklearn.model_selection import train_test_split as tts

In [9]:
xtrain,xtest,ytrain,ytest = tts(x,y,test_size=.3, random_state = 1)

In [10]:
xtrain.head()

Height  Female  Male
1277  72.385301       0     1
5361  60.960147       1     0
5728  69.119292       1     0
4276  68.271050       0     1
3112  64.991405       0     1

In [11]:
xtest.head()

Height  Female  Male
7154  65.068038       1     0
3258  70.824643       0     1
2188  68.663594       0     1
5609  65.987610       1     0
3186  70.901461       0     1

# Linear Regression

In [12]:
from sklearn.linear_model import LinearRegression

In [13]:
le =LinearRegression()  #create object

In [14]:
le.fit(xtrain,ytrain)

LinearRegression()

In [15]:
pred = le.predict([[70,0,1]]) #suppose,taken hieght 70, female =0, male=1 
pred

C:\Users\MD.NAIM HOSSAIN\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:420: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([192.61199108])

In [16]:
le.score(xtrain,ytrain)  #trining score

0.8972135459668117

In [17]:
le.score(xtest,ytest)  #testing score

0.9059959607091161

# MSE for testing

In [18]:
from sklearn.metrics import mean_squared_error

In [19]:
#testing mse
test_mse = mean_squared_error(ytest, le.predict(xtest))
test_mse

99.6725099807003

# KNN Regression

In [20]:
from sklearn.neighbors import KNeighborsRegressor

In [21]:
kr= KNeighborsRegressor()

In [22]:
kr.fit(xtrain,ytrain)

KNeighborsRegressor()

In [23]:
predict = kr.predict([[70,0,1]]) #suppose,taken hieght 70, female =0, male=1 
predict

C:\Users\MD.NAIM HOSSAIN\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:420: UserWarning: X does not have valid feature names, but KNeighborsRegressor was fitted with feature names
  warnings.warn(


array([196.72954164])

In [24]:
kr.score(xtrain,ytrain)  #trining score

0.917179497300704

In [25]:
kr.score(xtest , ytest)  #testing score

0.8838502073708883

# MSE for testing

In [26]:
#testing mse
test_mse = mean_squared_error(ytest, kr.predict(xtest))
test_mse

123.15365863436983

# Compare KNN and Linear Regression

KNN predict value is: 196.72954164         &   Linear Re predict value is: 192.61199108
KNN Training score is 0.917179497300704    &   LR Training score is 0.8972135459668117
KNN testing score is  0.8838502073708883   &   LR Testing score is  0.9059959607091161
KNN MSE value is      123.15365863436983   &   LR MSE value is      99.6725099807003


# Step 02: Classification

In [27]:
df2.head()

Gender     Height      Weight
0   Male  73.847017  241.893563
1   Male  68.781904  162.310473
2   Male  74.110105  212.740856
3   Male  71.730978  220.042470
4   Male  69.881796  206.349801

2. Separate x and (y=Gender)
3. Train = 70%, Test = 30%
4. Apply KNN Classifier 
5. Evaluate the Model by only Accuracy.
6. Apply KNN Classifier: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html

Work from here..

# separate X and Y

In [28]:
x=df2.drop(['Gender'],axis=1)
x.head()

Height      Weight
0  73.847017  241.893563
1  68.781904  162.310473
2  74.110105  212.740856
3  71.730978  220.042470
4  69.881796  206.349801

In [31]:
y=df2['Gender']
y.head()

0    Male
1    Male
2    Male
3    Male
4    Male
Name: Gender, dtype: object

# Training part

In [32]:
x_train,x_test,y_train,y_test=tts(x,y,test_size=.3,random_state=1)

In [33]:
x_train.head()

Height      Weight
1277  72.385301  189.165592
5361  60.960147  124.876796
5728  69.119292  168.864403
4276  68.271050  172.941248
3112  64.991405  144.551044

# KNN Classifier

In [34]:
from sklearn.neighbors import KNeighborsClassifier

In [35]:
kc =KNeighborsClassifier()

In [37]:
kc.fit(x_train,y_train)

KNeighborsClassifier()

# Accuracy

In [38]:
kc.score(x_train, y_train)

0.9263527054108216

In [39]:
kc.score(x_test,y_test)

0.9146864043630697

In [42]:
predict = kc.predict([[70.0,170.0]])
predict

C:\Users\MD.NAIM HOSSAIN\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:420: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


array(['Male'], dtype=object)